src/client/Microsoft.Identity.Client/Internal/Broker/BrokerUsernamePasswordRequestComponent.cs

```diff
@@ -0,0 +1,84 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using Microsoft.Identity.Client.ApiConfig.Parameters;
+using Microsoft.Identity.Client.Core;
+using Microsoft.Identity.Client.Internal.Requests;
+using Microsoft.Identity.Client.OAuth2;
+using Microsoft.Identity.Client.Utils;
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Identity.Client.Internal.Broker
+{
+
+    internal class BrokerUsernamePasswordRequestComponent : ITokenRequestComponent
+    {
+        internal IBroker Broker { get; }
+        private readonly AcquireTokenByUsernamePasswordParameters _usernamePasswordParameters;
+        private readonly string _optionalBrokerInstallUrl; // can be null
+        private readonly AuthenticationRequestParameters _authenticationRequestParameters;
+        private readonly IServiceBundle _serviceBundle;
+        private readonly ICoreLogger _logger;
+
+        public BrokerUsernamePasswordRequestComponent(
+            AuthenticationRequestParameters authenticationRequestParameters,
+            AcquireTokenByUsernamePasswordParameters acquireTokenByUsernamePasswordParameters,
+            IBroker broker)
+        {
+            _authenticationRequestParameters = authenticationRequestParameters;
+            _usernamePasswordParameters = acquireTokenByUsernamePasswordParameters;
+            _serviceBundle = authenticationRequestParameters.RequestContext.ServiceBundle;
+            Broker = broker;
+            _logger = _authenticationRequestParameters.RequestContext.Logger;
+        }
+
+        public async Task<MsalTokenResponse> FetchTokensAsync(CancellationToken cancellationToken)
+        {
+            if (Broker.IsBrokerInstalledAndInvokable(_authenticationRequestParameters.AuthorityInfo.AuthorityType))
```

Comment:
The fallback condition is not unit tested: if broker is not installed or available, MSAL should do ROPC on its own. This should be tested via unit tests, not integration test.